File: 01200. Minimum Absolute Difference/solution.go
/**
 * Problem: 1200. Minimum Absolute Difference
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 4 ms (Beats 100%)
 */

func minimumAbsDifference(arr []int) [][]int {
	// Initialize minVal and maxVal to track the minimum and maximum values in arr
	minVal, maxVal := math.MaxInt, math.MinInt

	// Find the minimum and maximum values in arr
	for _, v := range arr {
		minVal = min(minVal, v)
		maxVal = max(maxVal, v)
	}

	// Create a boolean slice to mark the existence of each value in arr
	exists := make([]bool, maxVal-minVal+1)

	// Mark the presence of each value in arr
	for _, v := range arr {
		exists[v-minVal] = true
	}

	// Initialize minDiff to the largest possible integer value
	minDiff := math.MaxInt
	var prev int // prev stores the previous index with a value present
	var result [][]int // result stores pairs with the minimum absolute difference

	// Iterate through the exists slice to find pairs with the minimum absolute difference
	for i := 1; i < len(exists); i++ {
		if exists[i] {
			diff := i - prev // Calculate the difference between current and previous value

			if diff == minDiff {
				// If current difference equals minDiff, append the pair to result
				result = append(result, []int{prev + minVal, i + minVal})
			} else if diff < minDiff {
				// If current difference is less than minDiff, update minDiff and reset result
				minDiff = diff
				result = [][]int{{prev + minVal, i + minVal}}
			}

			// Update prev to current index
			prev = i
		}
	}

	// Return all pairs with the minimum absolute difference
	return result
}

File: 01200. Minimum Absolute Difference/solution.cs
/**
 * Problem: 1200. Minimum Absolute Difference
 *
 * Difficulty: Easy
 *
 * Language: C#
 *
 * Performance: Runtime - 8 ms (Beats 100%)
 */

public class Solution {
    public IList<IList<int>> MinimumAbsDifference(int[] arr) {
        // Find minimum and maximum values in array
        var minVal = arr[0];
        var maxVal = arr[0];
        
        // Scan array to determine value range
        for (var i = 1; i < arr.Length; ++i) {
            var num = arr[i];
            
            minVal = Math.Min(minVal, num);
            maxVal = Math.Max(maxVal, num);
        }
        
        // Calculate maximum possible difference
        var minDifference = maxVal - minVal;
        
        // Create boolean array to mark which values exist
        var buckets = new bool[minDifference + 1];
        
        // Mark all present values in buckets array
        for (var i = 0; i < arr.Length; ++i)
            buckets[arr[i] - minVal] = true;
        
        // Initialize result list
        var res = new List<IList<int>>();
        
        // Track previous value for difference calculation
        var prevVal = minVal;
        
        // Scan through all possible values in range
        for (var i = 1; i < buckets.Length; ++i) {
            // Skip if value doesn't exist in input
            if (!buckets[i]) continue;
            
            // Calculate current value from index
            var num = minVal + i;
            
            // Calculate difference with previous value
            var diff = num - prevVal;
            var leftVal = prevVal;
            
            // Update previous value for next iteration
            prevVal = num;
            
            // Skip if difference is larger than current minimum
            if (diff > minDifference) continue;
            
            // Reset result if smaller difference found
            if (diff < minDifference) {
                minDifference = diff;
                res.Clear();
            }
            
            // Add pair to result list
            res.Add(new[] { leftVal, num });
        }
        
        // Return all pairs with minimum absolute difference
        return res;
    }
}
